src/DistributedComputations/distributed_transpose.jl

```diff
@@ -0,0 +1,113 @@
+using Oceananigans.Grids: architecture
+using Oceananigans.Architectures: on_architecture
+using KernelAbstractions: @index, @kernel
+using MPI: VBuffer, Alltoallv!
+
+# Transpose directions are assumed to work only in the following configuration
+# z -> y -> x -> y -> z
+# where z stands for z-complete data, y for y-complete data, and x for x-complete data
+# The initial field is always assumed to be in the z-complete configuration
+
+# Fallbacks for slab decompositions
+transpose_z_to_y!(::SlabYFields) = nothing
+transpose_y_to_z!(::SlabYFields) = nothing
+transpose_x_to_y!(::SlabXFields) = nothing
+transpose_y_to_x!(::SlabXFields) = nothing
+
+# Since z -> y -> x -> y -> z we only nedd to define the `pack` and `unpack` kernels
+# for the x and z configurations once, y requires two definitions depending on which
+# configuration it's interacting with
+
+@kernel function _pack_buffer_z!(yzbuff, zfield, N)
+    i, j, k = @index(Global, NTuple)
+    @inbounds yzbuff.send[j + N[2] * (i-1 + N[1] * (k-1))] = zfield[i, j, k]
```

Comment:
Can you add a comment that explains this line?

A small clarification is to use

```julia
Nx, Ny, Nz = N
```

Reply:
so the buffers are vectors and we are packing them from fields?

Reply:
`Alltoallv!` distributes a one-dimensional vector between different processors, so we pack the 3D field into a one-dimensional vector, redistribute it across processors with the new configuration and unpack it again into the 3D fields. I can write some docs for this. For the moment this distribution happens with chunks of same size being distributed between cores (the default option for `Alltoallv`) so we allow only grid-sizes that are integer-divisible by the number of ranks. In future, I am planning to pass to `Alltoallv` also the sizes of the chunks to be able to have different number of points on different processors and alleviate the constraints we have now.

Reply:
Just a comment like "this function packs a 3D field into a 1D send buffer" is enough

Reply:
but of course the more docs the better